Support raw identifiers (backtick-delimited identifiers containing non-identifier characters).

allevato opened the pull request:
Implementation of [SE-0451](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0451-escaped-identifiers.md).

allevato committed:
Add support for raw identifiers.

Raw identifiers are backtick-delimited identifiers that can contain any
non-identifier character other than the backtick itself, CR, LF, or other
non-printable ASCII code units, and which are also not composed entirely
of operator characters.

allevato committed:
Add a test importing a Clang module with non-identifier characters.

This requires no additional work beyond just parsing the identifier because Clang module maps already support such names in quotes.

allevato committed:
Allow module aliases to use escaped identifiers as the alias name.

The original module names themselves must still be valid unescaped identifiers; most of the serialization logic in the compiler depends on the name of a module matching its name on the file system, and it would be very complex to turn escaped identifiers into file-safe names.

allevato committed:
Allow module aliases to be expressed in the explicit Swift module map JSON file.

For build systems that already generate these files, it makes sense to include the aliases so that the map file serves as a comprehensive index of how the module inputs are referenced.

rintaro commented on include/swift/AST/Identifier.h:
```diff
@@ -101,19 +101,22 @@ class Identifier {
     return isOperatorSlow();
   }
 
+  /// Returns true if this identifier contains non-identifier characters and
+  /// must always be escaped with backticks, even in contexts were other
+  /// escaped identifiers could omit backticks (like keywords as argument
+  /// labels).
+  bool mustAlwaysBeEscaped() const;
```
Could you add some test cases for code completion checking  `mustAlwaysBeEscaped()` items are actually escaped in code completion suggestion?

allevato replied:
Thanks! Those did need to be fixed.

allevato commented:
This PR is ready except for the issue mentioned in the description of specialized C++ templates getting escaped by backticks (and now since the backticks are added to the mangling, it changes their mangled names as well). After the holidays I'll take another look at how to handle these.

allevato committed:
Ensure that `#fileID` wraps raw identifier module names in backticks.

`#fileID` never accounted for the possibility that someone one have
a module alias _itself_, so it always generated the module's real
(physical) name. This _technically_ changes the behavior of `#fileID`
for self-aliased modules, but since nobody would have ever had a reason
to do that before raw identifiers, it's unlikely that this change would
affect anyone in practice.

allevato committed:
[Sema] Validate names of `@objc` declarations with raw identifiers.

If a decl is exported to Objective-C (explicitly or implicitly), it
must be given an explicit name that is a valid Objective-C identifier.

allevato committed:
Preserve existing printing/mangling behavior for C++ specializations.

Imported C++ template specializations receive identifiers that contain
their type signature; e.g., `X<Y, Z>`. Since this means the identifier
contains non-identifier characters, the new behavior was trying to
escape them with backticks in ASTPrinter, ASTMangler, and the runtime
metadata. This pulls that back to preserve the current behavior for
specifically those types.

allevato commented:
The issues with imported C++ template specializations have been fixed.

allevato commented:
https://github.com/swiftlang/swift-syntax/pull/2857

@swift-ci please test

allevato commented:
@swift-ci please test

allevato commented:
https://github.com/swiftlang/swift-syntax/pull/2857

@swift-ci please test

grynspan commented:
Is there a named feature I can test for with `hasFeature()`? We may not be able to adopt without it because we risk breaking on older 6.2 toolchains that don't support raw identifiers.

allevato committed:
Add `RawIdentifiers` as a language feature.

This lets clients test `#if hasFeature(RawIdentifiers)` to
determine compiler support.

allevato commented:
> Is there a named feature I can test for with `hasFeature()`? We may not be able to adopt until 6.3 without it because we risk breaking on older 6.2 toolchains that don't support raw identifiers.

Looks like `LANGUAGE_FEATURE(RawIdentifiers, ...)` works for this, without being an upcoming or experimental feature (it's neither). Thanks for the suggestion!

allevato commented:
https://github.com/swiftlang/swift-syntax/pull/2857

@swift-ci please test

allevato commented:
https://github.com/swiftlang/swift-syntax/pull/2857

@swift-ci please test

allevato commented:
https://github.com/swiftlang/swift-syntax/pull/2857

@swift-ci please test

allevato commented:
https://github.com/swiftlang/swift-syntax/pull/2857

@swift-ci please test macOS platform

allevato commented:
https://github.com/swiftlang/swift-syntax/pull/2857

@swift-ci please test macOS platform

allevato commented:
@DougGregor @jckarter Can you review please, or forward along to someone who can? Thanks!

rintaro commented on lib/IDE/CompletionLookup.cpp:
```diff
@@ -476,7 +476,11 @@ void CompletionLookup::addValueBaseName(CodeCompletionResultBuilder &Builder,
+  if (Name.mustAlwaysBeEscaped()) {
+    // Names that are raw identifiers must always be escaped regardless of
+    // their position.
+    shouldEscapeKeywords = true;
```
This calls `Lexer:: identifierMustAlwaysBeEscaped(StringRef)` twice, in `Name.mustAlwaysBeEscaped()` and `Builder.escapeKeyword()` which is a waste. Can we avoid that? I'm fine with just manually wrapping it with back-ticks here, or you could factor out the escaping logic from  `Builder.escapeKeyword`.

allevato replied:
Done. I've factored out a `escapeWithBackticks` method in the builder so it can be called directly.

rintaro commented on lib/IDE/CompletionLookup.cpp:
```diff
@@ -498,6 +502,16 @@ void CompletionLookup::addValueBaseName(CodeCompletionResultBuilder &Builder,
+                                     Identifier Name) {
+  if (Name.mustAlwaysBeEscaped()) {
+    SmallString<16> buffer;
+    Builder.addBaseName(Builder.escapeKeyword(Name.str(), true, buffer));
```
Ditto

allevato replied:
Done.

rintaro commented on test/SourceKit/CodeComplete/complete_raw_identifiers.swift:
Could you use `swift-ide-test`  in `test/IDE` for testing these? Our primary place for testing code completion results is `test/IDE/complete_*.swift`.
`test/SourceKit/CodeComplete` is for SourceKit specific functionalities, like response format, ordering or such.

allevato replied:
Moved the tests to `test/IDE` and updated them to use `%batch-code-completion`.

rintaro reviewed (commented):
Parser and Index related changes looks good to me.
Deferring to others for other areas.

rintaro commented on lib/IDE/CodeCompletionResultBuilder.h:
```diff
@@ -382,18 +391,16 @@ class CodeCompletionResultBuilder {
       shouldEscape = llvm::StringSwitch<bool>(Word)
 #include "swift/AST/TokenKinds.def"
-        .Default(false);
+                         .Default(Lexer::identifierMustAlwaysBeEscaped(Word));
```
Nit: indentation

allevato replied:
This is what clang-format wants, apparently, because it's treated as a continuation of the expression `llvm::StringSwitch...` above.

rintaro reviewed (commented):
Parse/IDE/Index changes lgtm

DougGregor commented:
@allevato  I'm so sorry, I thought this review had been picked up before.

DougGregor commented on lib/AST/ASTMangler.cpp:
```diff
@@ -3062,7 +3062,8 @@ void ASTMangler::appendAnyGenericType(const GenericTypeDecl *decl,
       // name of the Swift decl which does include template arguments.
-      appendIdentifier(nominal->getName().str());
+      appendIdentifier(nominal->getName().str(),
+                       /*allowRawIdentifiers=*/false);
```
I feel like we should revisit the naming of imported class template specializations in light of raw identifiers, but I don't want to bring any of that discussion into this pull request.

DougGregor commented on lib/Index/Index.cpp:
```diff
@@ -913,10 +913,18 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+        assert(Range.getByteLength() > 1);
+        if (Range.str().front() == '`') {
+          assert(Range.getByteLength() > 3);
+          assert(Range.str().consume_front("`_") ||
+                 Range.str().consume_front("`$"));
```
I think you want `front` rather than `consume_front`. It's just mutating the temporary, so it's not going to cause a problem despite being in the `assert`.

DougGregor commented on lib/Sema/TypeCheckDeclObjC.cpp:
```diff
@@ -1745,13 +1801,17 @@ bool IsObjCRequest::evaluate(Evaluator &evaluator, ValueDecl *VD) const {
-    if (auto attr = proto->getAttrs().getAttribute<ObjCAttr>()) {
+    auto attr = proto->getAttrs().getAttribute<ObjCAttr>();
+    if (attr) {
+      checkObjCNameValidity(proto, attr);
```
This `checkObjCNameValidity` should be outside of the `if` so that we diagnose, e.g.,

```swift
@objc
protocol `some protocol` { }
```

right?

[EDIT: No, I'm wrong, because you have a test like this. I'm not sure how it works, though]

DougGregor commented on test/Parse/escaped_identifiers.swift:
```diff
@@ -23,3 +23,42 @@ var applyGet: Int {
+func `+`(lhs: Int, rhs: Int) -> Int  // expected-error{{expected identifier in function declaration}}
+
+@propertyWrapper
+struct `@PoorlyNamedWrapper`<`The Value`> {
```
It's hideous! But makes a nice test case, thanks

DougGregor reviewed (approved):
This is looking great, please go ahead and merge

allevato committed:
Final tweaks from review comments.

allevato commented:
@swift-ci please test

allevato commented:
@swift-ci please test

allevato commented:
@swift-ci please test macOS platform
